bluepysnap/edges.py

```diff
@@ -480,7 +482,33 @@ def _get(self, selection, properties=None):
 
         return result
 
-    def ids(self, group=None, limit=None, sample=None):
+    def _edge_ids_by_filter(self, queries, raise_missing_prop):
+        """Return edge IDs if their properties match the `queries` dict.
+
+        `props` values could be:
+            pairs (range match for floating dtype fields)
+            scalar or iterables (exact or "one of" match for other fields)
+
+        You can use the special operators '$or' and '$and' also to combine different queries
+        together.
+
+        Examples:
+            >>> self._edge_ids_by_filter({ Edge.POST_SECTION_ID: (0, 1),
+            >>>                            Edge.AXONAL_DELAY: (.5, 2.) })
+            >>> self._edge_ids_by_filter({'$or': [{ Edge.PRE_X_CENTER: [2, 3]},
+            >>>                              { Edge.POST_SECTION_POS: (0, 1),
+            >>>                              Edge.SYN_WEIGHT: (0.,1.4) }]})
+
+        """
+        properties = query.get_properties(queries)
+        unknown_props = properties - self.property_names
+        if raise_missing_prop and unknown_props:
+            raise BluepySnapError(f"Unknown edge properties: {unknown_props}")
+        data = self.get(None, properties - unknown_props)
```

Comment:
I am quite scared this cannot be done. I allowed he None for the get because user is in full control but I am sure for huge circuit this will not be possible... Did you try ?

Reply:
> I am quite scared this cannot be done.

Should I understand this as "I am quite scared that this can be done"?

> I allowed he None

Should I understand this as "I allowed the `None`"?

> Did you try ?

No but I agree that this would highly likely fail/stuck on edges of Tb size with many properties.

1. In general this should work when there are only several properties within the query. All group attributes in edges files are int or floats. Lets assume float64 => 8 bytes. Lets say we have 1e9 (billion) edges in one population. One property then would take 8Gb. I doubt that edges of that size will be processed on a desktop computer. So, we have a high performant server hardware. Even for 10 properties it will take 80Gb of Ram which is not much for server hardware.

2. I don't want to introduce Dask until we are sure that libsonata will not support it for 100%.

Reply:
I was not clear enough. 

What I meant is: 
Back in the days, the edges "queries" : `get`, `afferent_nodes`, `efferent_nodes` forced as entries a list of edge ids. Then I chose to allow a None list of edge ids which means : take all the edge_ids.
I did that because this is not hidden to the user and he clearly needs to call it:
Ex :
```
get(None)
```
And the None is never a default value.

Using the queries I am a bit scared this is a bit more hidden and unclear why one fills his memory. 

I totally agree with the hardware but maybe we can do it a bit differently without dask. 
What about doing the property queries by chunk ? 

Something like : 
```
res = []
ids = self.ids(None)
critical_size = 60000000  # can be something computed using the dtypes of the properties
for chunk in np.array_split(ids, len(ids) // critical_size):
    data = self.get(chunk, properties - unknown_props)
    res.extend(query.resolve_ids(data, self.name, queries))
```
? 
doing so you reduce a lot the memory consumption because you don t load everything in memory at the same time and I guess with the change I made in #128 you should not have problems with the positional queries.

Reply:
Great idea with chunking, I like it. Let see how it behaves. The current approach failed for 1.6Tb files (9.12 billion edges) with option `--mem=100g` and a single property. Around 74Gb is needed for single property in this case. The error was expected: `Out Of Memory`.

Reply:
Thanks a lot for the check !
We will need to find a cleaver way of doing this so it can be usable ... I hope chunking can help.

Reply:
`--mem=0` worked for this 1.6Tb case but it took `1109.78255582` seconds.

Reply:
yeah I did not expect insane perfs anyway.
Maybe we will be able to improve afterwards with multi processing.

Reply:
ok, chunking is working and helps. It worked for 1.6Tb with `--mem=200gb`. Ran actually faster for `799.776455279999` seconds